File: src/civitas/MazoSorpresas.java
/*
 * To change this license header, choose License Headers in Project Properties.
 * To change this template file, choose Tools | Templates
 * and open the template in the editor.
 */
package civitas;
import java.util.ArrayList;
import java.util.Collections;

/**
 *
 * @author alejandrof
 */
public class MazoSorpresas {
    private ArrayList<Sorpresa> sorpresas;
    private Boolean barajada;
    private int usadas;
    private Boolean debug;
    private ArrayList<Sorpresa> cartasEspeciales;
    private Sorpresa ultimaSorpresa;
    
    private void init(){
        sorpresas = new ArrayList<>();
        cartasEspeciales = new ArrayList<>();
        barajada = false;
        usadas = 0;
    }
    
    MazoSorpresas(Boolean debugging){
        debug = debugging;
        this.init();
        if(debug)
            Diario.getInstance().ocurreEvento("Modo debug del Mazo activado");
    }
    
   MazoSorpresas(){
        debug = false;
        this.init();
    }
    
   void alMazo(Sorpresa s){
        if(!barajada)
            sorpresas.add(s);
    }
    
    Sorpresa siguiente(){
        if(!debug && (!barajada || usadas >= sorpresas.size())){
            Collections.shuffle(sorpresas);
            barajada = true;
            usadas = 0;
        }
        
        usadas++;
        Sorpresa extraida = sorpresas.get(0);
        sorpresas.add(extraida);
        sorpresas.remove(0);
        ultimaSorpresa = extraida;
        return extraida;
    }
    
    void inhabilitarCartaEspecial(Sorpresa sorpresa){
        Boolean remove = sorpresas.remove(sorpresa);
        
        if(remove){
            cartasEspeciales.add(sorpresa);
            Diario.getInstance().ocurreEvento("Se inhabilita la carta especial " + sorpresa.toString());
        }
    }
    
    void habilitarCartaEspecial(Sorpresa sorpresa){
        Boolean remove = cartasEspeciales.remove(sorpresa);
        
        if(remove){
            sorpresas.add(sorpresa);
            Diario.getInstance().ocurreEvento("Se habilita la carta especial " + sorpresa.toString());
        }
    }
    
    Sorpresa getUltimaSorpresa(){
        return ultimaSorpresa;
    }
}

File: Civitas_Ruby/lib/mazo_sorpresas.rb
#encoding:utf-8
# To change this license header, choose License Headers in Project Properties.
# To change this template file, choose Tools | Templates
# and open the template in the editor.

require_relative 'sorpresa.rb'
require_relative 'diario.rb'

module Civitas
  class MazoSorpresas
    
    attr_reader :ultimaSorpresa

    def init()
      @sorpresas = Array.new()
      @cartasEspeciales = Array.new()
      @barajada = false;
      @usadas = 0;
    end
    
    def initialize(modo = false)
      init()
      @debug = modo;
      if @debug then 
        Diario.instance().ocurre_evento("Modo debug del Mazo activado")
      end
    end
    
    def alMazo(s)
      if !@barajada then
        @sorpresas.push(s)
      end
    end
    
    def siguiente()
      if !@debug && (@usadas >= @sorpresas.count() || !@barajada) then
        @sorpresas.shuffle
        @barajada = true
        @usadas = 0
      end
      
      @usadas = @usadas + 1
      extraida = @sorpresas[0]
      @sorpresas.delete_at(0)
      @sorpresas.push(extraida)
      @ultimaSorpresa = extraida
      return extraida
    end
    
    def inhabilitarCartaEspecial(sorpresa)
      remove = nil
      remove = @sorpresas.delete(sorpresa)
      if remove != nil then
        @cartasEspeciales.push(sorpresa)
        Diario.instance.ocurre_evento("Se inhabilita la carta especial" + sorpresa.to_s())
        return true
      else
        return false
      end
    end
    
    def habilitarCartaEspecial(sorpresa)
      remove = nil
      remove = @cartasEspeciales.delete(sorpresa)
      if remove != nil then
        @sorpresas.push(sorpresa)
        Diario.instance.ocurre_evento("Se habilita la carta especial " + sorpresa.to_s())
        return true
      else
        return false
      end
    end
      
    private :init
  end
end
